### DISCUSSION TITLE
RuntimeError during unit tests: Future attached to a different loop

### DISCUSSION THREAD

**Discussion by prashantgupta24** [fastapi/fastapi#10518] [Questions]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from httpx import AsyncClient
from fastapi import FastAPI
from sse_starlette.sse import EventSourceResponse
import asyncio

app = FastAPI()

@app.post("/foo")
async def http_foo() -> EventSourceResponse:
    async def event_generator():
        yield {"data": "1"}
        yield {"data": "2"}
        yield {"data": "3"}
    return EventSourceResponse(event_generator())

def parse_event_stream(text):
    events = []
    for line in text.strip().split("\r\n\r\n"):
        events.append(line[len("data:"):].strip())
    return events

async def test_first():
    client = AsyncClient(app=app, base_url="http://test")
    async with client:
        response = await client.post("/foo")
    events = parse_event_stream(response.text)
    assert events == ["1", "2", "3"]

async def test_second():
    client = AsyncClient(app=app, base_url="http://test")
    async with client:
        response = await client.post("/foo")
    events = parse_event_stream(response.text)
    assert events == ["1", "2", "3"]
```

### Description

Quoting from here: https://github.com/sysid/sse-starlette/issues/68. Asking question here since no response in `sse_starlette`:

```I tried to write pytest testcases for a FastAPI route that returns an event source/stream. The first test always succeedes, but the second test, despite identical content, consistently runs into an error: RuntimeError: Task ... got Future ... attached to a different loop. I'm not sure if the error is related to sse_starlette, fastapi, httpx or pytest-asyncio.```

### Operating System

macOS

### FastAPI Version

0.104.0

### Pydantic Version

2.4.2

### Python Version

3.9.2

**Comment by ArnaudPel** [comment]
In case that helps, I managed to make the example code above work simply by decorating both test functions as follows:

```
@pytest.mark.asyncio(scope="session")
async def test_first():
    ...

@pytest.mark.asyncio(scope="session")
async def test_second():
    ...
```

- fastapi==0.115.4
- httpx==0.27.2
- pytest==8.3.3
- pytest-asyncio==0.24.0
- sse-starlette==2.1.3

**Comment by ya-b** [comment]
I had the same issue, and it worked by changing EventSourceResponse to StreamingResponse.
```
from fastapi.responses import StreamingResponse

@app.post("/foo")
async def http_foo() -> StreamingResponse:
    async def event_generator():
        str_data = 'xxxxxxxxx'
        yield f'data: {str_data }\n\n'
    return StreamingResponse(event_generator())
```

**Comment by YuriiMotov** [comment]
Let's close this for now to keep the discussions clean and organized.
If you’ve found a solution, please share it here.
If you still need help, feel free to let us know.
